tests/test-MPIcomm.C

```diff
@@ -0,0 +1,448 @@
+    { 'i', "-i I", "Set the number of times to do the random unit tests.", TYPE_INT,     &niter },
+    END_OF_ARGUMENTS
+  };	
+  parseArguments (argc, argv, args); 
+#ifdef __LINBOX_HAVE_MPI
+  MPI_Bcast(&ni, 1, MPI_INT, 0, MPI_COMM_WORLD); 
+  MPI_Bcast(&niter, 1, MPI_INT, 0, MPI_COMM_WORLD);
+#else
+  srand (time(NULL));
+#endif
+  nj=ni;
+    
+  Givaro::ZRing<float> ZZ;
+//  DenseMatrix<Givaro::ZRing<Integer> > A (ZZ,ni,nj);
+SparseMatrix<Givaro::ZRing<float>> A (ZZ,ni,nj);
+  
+  typedef BlasVector<Givaro::ZRing<float> > DenseVector;
+  DenseVector X(ZZ, A.rowdim()), X2(ZZ, A.rowdim()),  B(ZZ, A.rowdim());
+    
+  for(int j=0;j<niter;j++){  
+    
+#ifdef __LINBOX_HAVE_MPI  
+    if(0==Cptr->rank()){
+#endif
+      genData (A, bits);
+      genData (B, bits);
+      
+      //LinBox::rank (r, A); std::cout<<"The rank of generated matrix A is:"<<r<<std::endl;  
+      
+      
+	std::cerr << ">>>>Compute with B: " << std::endl;      
+	for(int j=0;j<nj;j++) std::cerr << B.getEntry(j) << std::endl; 
+	
+	std::cout << "Compute with A: " << A.rowdim() << " by " << A.coldim() << std::endl;
+	if (A.rowdim() <= 20 && A.coldim() <= 20) A.write(std::cout << "A:=",Tag::FileFormat::Maple) << ';' << std::endl;
+        
+#ifdef __LINBOX_HAVE_MPI 	
+    }//End of BLock for process(0)
+#endif
+  
+#ifdef __LINBOX_HAVE_MPI
+    //distribute big integer compatible data
+    {
+      //double starttime, endtime; 
+      //MPI_Barrier(MPI_COMM_WORLD);
+      //starttime = MPI_Wtime();
+      //MPI data distribution for Integer type value
+//      Cptr->bcast(A,0);   // MPIgmpBcast(A, ni, nj, 0, Cptr);
+//      Cptr->bcast(B,0);   // MPIgmpBcast(B, ni, 0, Cptr);
+if(0==Cptr->rank())      Cptr->send(A,1);   // MPIgmpBcast(A, ni, nj, 0, Cptr);
+if(0==Cptr->rank())      Cptr->send(B,1);   // MPIgmpBcast(B, ni, 0, Cptr);
+
+if(0!=Cptr->rank())      Cptr->recv(A,0);   // MPIgmpBcast(A, ni, nj, 0, Cptr);
+if(0!=Cptr->rank())      Cptr->recv(B,0);   // MPIgmpBcast(B, ni, 0, Cptr);
+      //MPI_Barrier(MPI_COMM_WORLD);
+      //endtime   = MPI_Wtime(); 
+      //std::cout<<"MPI data distribution used CPU time (seconds): " <<endtime-starttime<<std::endl;
+    }
+#endif
+
+   
+    //Check if data are correctly distributed to all processes
+    
+      if(0!=Cptr->rank()){std::cerr <<"process("<<Cptr->rank()<< ")<<<<Compute with B: " << std::endl;
+      for(int j=0;j<nj;j++) std::cerr << B.getEntry(j) << std::endl; 
+      
+      std::cerr <<"process("<<Cptr->rank()<< ")<<<<Compute with A: " << A.rowdim() << " by " << A.coldim() << std::endl;
+      if (A.rowdim() <= 20 && A.coldim() <= 20) A.write(std::cout << "A:=",Tag::FileFormat::Maple) << ';' << std::endl;}
+  
+/*    
+    if(!test_set(X2, A, B
+#ifdef __LINBOX_HAVE_MPI
+		 , Cptr
+#endif
+		 )) break;
+*/
+  }
+
+#ifdef __LINBOX_HAVE_MPI
+  if(0 == Cptr->rank()){  
+#endif
+#ifdef __LINBOX_HAVE_MPI    
+    MPI_Finalize();
+#endif
+    return 0;
+#ifdef __LINBOX_HAVE_MPI
+  }
+  MPI_Finalize();
+#endif
```

Comment:
Seems like it can be simplified to:
```
#ifdef __LINBOX_HAVE_MPI
MPI_Finalize();
#endif

return 0;
```
(The `if` doesn't seem to do much, as we still call MPI_Finalize afterwards.)

Reply:
The process 0 will call the MPI_Finalize() then return with 0 and the other processes will only call MPI_Finalize() without return, which means once the process 0 invokes the return instruction it will not execute the second MPI_Finalize() instruction. I agree that it is possible to restructure the code in a different form so if you have any more elegant way to restructure, all suggestions are welcome.

Reply:
Is this required by anything to specifically not return a value?
Because C++ spec says that ending `main()` without returning a value is equivalent to do a `return 0;`. (https://stackoverflow.com/a/19293663)

Reply:
The MPI will let all processes running since the beginning of the whole program, only process 0 will return and it receive results from other processes and checks it meanwhile other processes may be still running, the instruction MPI_Finalize only suggests to terminate the process and it lets the MPI to decide. So if the whole program is invoked inside a loop then other processes will not return but the process 0 will return and check results.